### PR TITLE
jwt complete

### PR DESCRIPTION
**Pull request by rahulmalakar42**
- secured member routes for only logged in users
- password hasing using bcrypt
- cookie parsing
- error handling for jwt verification
- checking user jwt for every get request
- logout function

### CODE REVIEW

File: backend/app.js [hunk 1]
```diff
@@ -25,15 +30,33 @@ app.use(express.static(path.join(__dirname, '..', 'frontend', 'public')));
...
 app.use(morgan('dev'));
+app.use(cookieParser());
 
 //base url
+app.get('*', checkUser)
 app.get('/', (req, res) => {
   res.redirect('/books');
 });
 
+//cookies
```
Comment (noobg1) [inline backend/app.js:41]: lets remove the unused code?

File: backend/controllers/authcontroller.js [hunk 2]
```diff
@@ -1,15 +1,49 @@
...
+    if(err.message === 'incorrect email'){
+        errors.email = 'email is not registered'
+    }
+    //incorrect password
+    if(err.message === 'incorrect password'){
+        errors.password = 'password is incorrect'
+    }
+
+    //duplicate
+    if(err.code===11000){
```
Comment (noobg1) [inline backend/controllers/authcontroller.js:20]: what's this 11000?

File: backend/middleware/authmiddleware.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+const jwt = require('jsonwebtoken');
+const User = require('../models/user');
+
+const requireAuth = (req,res, next)=>{
+    const token = req.cookies.jwt;
+    //check if token exist and is verified
+    if(token){
+        jwt.verify(token, 'rahul secret', (err, decodedToken)=>{
```
Comment (noobg1) [inline backend/middleware/authmiddleware.js:8]: move the secret to environment variables?

File: backend/middleware/authmiddleware.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    if(token){
+        jwt.verify(token, 'rahul secret', async (err, decodedToken)=>{
+            if(err){
+                console.log(err.message)
+                res.locals.user = null;
+                next();
+            }
+            else{
+                console.log(decodedToken);
+                let user = await User.findById(decodedToken.id);
```
Comment (noobg1) [inline backend/middleware/authmiddleware.js:36]: any reason for `let` usage instead of const?

Comment (rahulmalakar42) [reply backend/middleware/authmiddleware.js]: I was defining it previously and assigning a new value in else block. Forgot to change to const afterwards. Resolving it now.

File: backend/controllers/authcontroller.js (hunk 2 above)
Comment (rahulmalakar42) [reply backend/controllers/authcontroller.js]: if account already exist and user signs up again with same credential, this comes up to be the error. I have resolved it now by using a variable for this in status code file.

File: backend/middleware/authmiddleware.js (hunk 3 above)
Comment (rahulmalakar42) [reply backend/middleware/authmiddleware.js]: Done.

**Pull request merged**